### ISSUE TITLE
ffmpeg producing an out-of-sync audio/video file if you download separate "best" video and audio files ... 

### ISSUE THREAD

**Issue by Albretch** [yt-dlp/yt-dlp#6674]
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm asking a question and **not** reporting a bug or requesting a feature
- [X] I've looked through the [README](https://github.com/yt-dlp/yt-dlp#readme)
- [X] I've verified that I'm running yt-dlp version **2023.03.04** ([update instructions](https://github.com/yt-dlp/yt-dlp#update)) or later (specify commit)
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar questions **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)

### Please make sure the question is worded well enough to be understood

 using the the format option: --format "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best"

 supposedly downloads the best video and audio for an mp4 video type if it exists otherwise the best other option, but it needs then ffmpeg to consolidate both files into a single video which is then out-of-sync! I have noticed other people complaining about it:

 https://stackoverflow.com/questions/8844460/ffmpeg-is-generating-an-out-of-sync-audio-video-file

 given the format options you have, for example:

$ ./yt-dlp --list-formats yDNVPDwrKmE
[youtube] Extracting URL: yDNVPDwrKmE
[youtube] yDNVPDwrKmE: Downloading webpage
[youtube] yDNVPDwrKmE: Downloading android player API JSON
[info] Available formats for yDNVPDwrKmE:
ID  EXT   RESOLUTION FPS CH │   FILESIZE   TBR PROTO │ VCODEC          VBR ACODEC      ABR ASR MORE INFO
──────────────────────────────────────────────────────────────────────────────────────────────────────────────────
sb2 mhtml 48x27        0    │                  mhtml │ images                                  storyboard
sb1 mhtml 80x45        0    │                  mhtml │ images                                  storyboard
sb0 mhtml 160x90       0    │                  mhtml │ images                                  storyboard
599 m4a   audio only      2 │    2.08MiB   31k dash  │ audio only          mp4a.40.5   31k 22k ultralow, m4a_dash
600 webm  audio only      2 │    2.32MiB   34k dash  │ audio only          opus        34k 48k ultralow, webm_dash
139 m4a   audio only      2 │    3.29MiB   49k dash  │ audio only          mp4a.40.5   49k 22k low, m4a_dash
249 webm  audio only      2 │    3.24MiB   48k dash  │ audio only          opus        48k 48k low, webm_dash
250 webm  audio only      2 │    4.08MiB   61k dash  │ audio only          opus        61k 48k low, webm_dash
140 m4a   audio only      2 │    8.73MiB  129k dash  │ audio only          mp4a.40.2  129k 44k medium, m4a_dash
251 webm  audio only      2 │    7.70MiB  114k dash  │ audio only          opus       114k 48k medium, webm_dash
17  3gp   176x144     12  1 │    5.38MiB   80k https │ mp4v.20.3       80k mp4a.40.2    0k 22k 144p
597 mp4   256x144     12    │    2.14MiB   32k dash  │ avc1.4d400b     32k video only          144p, mp4_dash
598 webm  256x144     12    │    2.07MiB   31k dash  │ vp9             31k video only          144p, webm_dash
394 mp4   256x144     24    │    4.84MiB   72k dash  │ av01.0.00M.08   72k video only          144p, mp4_dash
160 mp4   256x144     24    │    6.50MiB   97k dash  │ avc1.4d400c     97k video only          144p, mp4_dash
278 webm  256x144     24    │    6.28MiB   93k dash  │ vp9             93k video only          144p, webm_dash
395 mp4   426x240     24    │   10.22MiB  152k dash  │ av01.0.00M.08  152k video only          240p, mp4_dash
133 mp4   426x240     24    │   13.28MiB  197k dash  │ avc1.4d4015    197k video only          240p, mp4_dash
242 webm  426x240     24    │   13.67MiB  203k dash  │ vp9            203k video only          240p, webm_dash
396 mp4   640x360     24    │   20.91MiB  310k dash  │ av01.0.01M.08  310k video only          360p, mp4_dash
134 mp4   640x360     24    │   25.35MiB  376k dash  │ avc1.4d401e    376k video only          360p, mp4_dash
18  mp4   640x360     24  2 │ ~ 34.85MiB  505k https │ avc1.42001E    505k mp4a.40.2    0k 44k 360p
243 webm  640x360     24    │   25.63MiB  380k dash  │ vp9            380k video only          360p, webm_dash
397 mp4   854x480     24    │   36.77MiB  546k dash  │ av01.0.04M.08  546k video only          480p, mp4_dash
135 mp4   854x480     24    │   38.53MiB  572k dash  │ avc1.4d401e    572k video only          480p, mp4_dash
244 webm  854x480     24    │   46.69MiB  693k dash  │ vp9            693k video only          480p, webm_dash
22  mp4   1280x720    24  2 │ ~ 65.80MiB  954k https │ avc1.64001F    954k mp4a.40.2    0k 44k 720p
398 mp4   1280x720    24    │   65.89MiB  978k dash  │ av01.0.05M.08  978k video only          720p, mp4_dash
136 mp4   1280x720    24    │   55.58MiB  825k dash  │ avc1.4d401f    825k video only          720p, mp4_dash
247 webm  1280x720    24    │   85.08MiB 1263k dash  │ vp9           1263k video only          720p, webm_dash
399 mp4   1920x1080   24    │  111.10MiB 1649k dash  │ av01.0.08M.08 1649k video only          1080p, mp4_dash
137 mp4   1920x1080   24    │  173.31MiB 2572k dash  │ avc1.640028   2572k video only          1080p, mp4_dash
248 webm  1920x1080   24    │  137.41MiB 2039k dash  │ vp9           2039k video only          1080p, webm_dash
~
 What is the format option to download the best video and audio pair without having to use ffmpeg?

 How do you use that format option with yt-dlp?

 lbrtchx

### Provide verbose output that clearly demonstrates the problem

- [ ] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [ ] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

**Comment by bashonly** [comment]
> using the the format option: --format "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best"
> supposedly downloads the best video and audio for an mp4 video type if it exists otherwise the best other option, but it needs then ffmpeg to consolidate both files into a single video which is then out-of-sync!

I can't reproduce your issue. I ran this:
```
yt-dlp  --format "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best" yDNVPDwrKmE
```
and the resulting merged mp4 video had no video/audio sync issues.
What version of ffmpeg are you using? Show a verbose log.

> I have noticed other people complaining about it:
> https://stackoverflow.com/questions/8844460/ffmpeg-is-generating-an-out-of-sync-audio-video-file

This question is about re-encoding from flv to h264/aac and doesn't really have anything to do with the `-c copy` merging that ffmpeg performs under yt-dlp

> What is the format option to download the best video and audio pair without having to use ffmpeg?

`-f bv,ba`
but this will download the video and audio separately and not merge them. yt-dlp can only merge with ffmpeg

`-f b` will download the best quality format that is video+audio combined, however with youtube videos this means 720p

**Comment by Albretch** [comment]
> What version of ffmpeg are you using?
$ ffmpeg --version 2>&1 | head -n 1
ffmpeg version 4.3.5-0+deb11u1 Copyright (c) 2000-2022 the FFmpeg developers

> Show a verbose log.
$ ./yt-dlp --verbose --format "bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best" yDNVPDwrKmE
[debug] Command-line config: ['--verbose', '--format', 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best', 'yDNVPDwrKmE']
[debug] Encodings: locale UTF-8, fs utf-8, pref UTF-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2023.03.04 [392389b7d] (zip)
[debug] Python 3.9.2 (CPython x86_64 64bit) - Linux-5.10.0-18-amd64-x86_64-with-glibc2.31 (OpenSSL 1.1.1n  15 Mar 2022, glibc 2.31)
[debug] exe versions: ffmpeg 4.3.5-0, ffprobe 4.3.5-0
[debug] Optional libraries: certifi-2020.06.20, mutagen-1.45.1, no_Cryptodome-None, sqlite3-2.6.0
[debug] Proxy map: {}
[debug] Loaded 1786 extractors
[youtube] Extracting URL: yDNVPDwrKmE
[youtube] yDNVPDwrKmE: Downloading webpage
[youtube] yDNVPDwrKmE: Downloading android player API JSON
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, vcodec:vp9.2, channels, acodec, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, hdr:12(7), source, vcodec:vp9.2(10), channels, acodec, lang, proto, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, id
[info] yDNVPDwrKmE: Downloading 1 format(s): 137+140
[debug] Invoking dashsegments downloader on "https://rr2---sn-vgqsknsk.googlevideo.com/videoplayback?expire=1680156883&ei=c9QkZKGEFKiNtQefgKyICA&ip=35.130.58.74&id=o-AHULEq3KQhLivrkV1bqvy6gvPDDYUo26JnTaxEV0Sin-&itag=137&source=youtube&requiressl=yes&mh=x3&mm=31%2C26&mn=sn-vgqsknsk%2Csn-q4fzen7e&ms=au%2Conr&mv=m&mvi=2&pl=19&initcwndbps=1517500&spc=99c5CbDPcIfWe7detKZmjIHRFtuGonQ&vprv=1&svpuc=1&mime=video%2Fmp4&gir=yes&clen=181730194&dur=565.273&lmt=1610144438403862&mt=1680134957&fvip=3&keepalive=yes&fexp=24007246&c=ANDROID&txp=5535432&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cspc%2Cvprv%2Csvpuc%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIhAP1aKipKVCeQhojyUQjqWOD0UAP8orv9ZtkkhfGUAIkRAiB4S3fRbvAW-CkGkwXtYyviGlZ_6nSA7gZtOhesEPJYzg%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhANWS5pGBdk43bcYRRL7jZy4RfxWRY1FBFXqiAmlACDCLAiAnR1wjksTRACfs3_VsX4bJSIiG7Opj9FQVQk-NN8xTwQ%3D%3D"
[dashsegments] Total fragments: 18
[download] Destination: Types of Commercials [yDNVPDwrKmE].f137.mp4
[download] 100% of  173.31MiB in 00:00:18 at 9.15MiB/s
[debug] Invoking dashsegments downloader on "https://rr2---sn-vgqsknsk.googlevideo.com/videoplayback?expire=1680156883&ei=c9QkZKGEFKiNtQefgKyICA&ip=35.130.58.74&id=o-AHULEq3KQhLivrkV1bqvy6gvPDDYUo26JnTaxEV0Sin-&itag=140&source=youtube&requiressl=yes&mh=x3&mm=31%2C26&mn=sn-vgqsknsk%2Csn-q4fzen7e&ms=au%2Conr&mv=m&mvi=2&pl=19&initcwndbps=1517500&spc=99c5CbDPcIfWe7detKZmjIHRFtuGonQ&vprv=1&svpuc=1&mime=audio%2Fmp4&gir=yes&clen=9150638&dur=565.336&lmt=1610142364622764&mt=1680134957&fvip=3&keepalive=yes&fexp=24007246&c=ANDROID&txp=5531432&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cspc%2Cvprv%2Csvpuc%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgC2iqZVi9zNYu1zG8Gxy7c61iRPY0wd8CSChpo_4vWakCIG_LzFid2pnIe21y95ilrlFJprbnnKMhwAS6ak3cJg1p&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhANWS5pGBdk43bcYRRL7jZy4RfxWRY1FBFXqiAmlACDCLAiAnR1wjksTRACfs3_VsX4bJSIiG7Opj9FQVQk-NN8xTwQ%3D%3D"
[dashsegments] Total fragments: 1
[download] Destination: Types of Commercials [yDNVPDwrKmE].f140.m4a
[download] 100% of    8.73MiB in 00:00:00 at 8.92MiB/s
[Merger] Merging formats into "Types of Commercials [yDNVPDwrKmE].mp4"
[debug] ffmpeg command line: ffmpeg -y -loglevel repeat+info -i 'file:Types of Commercials [yDNVPDwrKmE].f137.mp4' -i 'file:Types of Commercials [yDNVPDwrKmE].f140.m4a' -c copy -map 0:v:0 -map 1:a:0 -movflags +faststart 'file:Types of Commercials [yDNVPDwrKmE].temp.mp4'
Deleting original file Types of Commercials [yDNVPDwrKmE].f137.mp4 (pass -k to keep)
Deleting original file Types of Commercials [yDNVPDwrKmE].f140.m4a (pass -k to keep)
$ 

> -f b will download the best quality format that is video+audio combined, however with youtube videos this means 720p
 Is there a way to fix the ffmpeg problem?

 Actually, I have no idea exactly why I don't notice the out-of-sync audio/video problem now.

 Do you notice anything fishy in the verbose log?

**Comment by bashonly** [comment]
> Actually, I have no idea exactly why I don't notice the out-of-sync audio/video problem now.

do you mean it's not out-of-sync now?

> Do you notice anything fishy in the verbose log?

looks OK

**Comment by Albretch** [comment]
OK, I think I figured out why I was getting those out-of-sync videos. I go into exposed mode (go online) using a Debian live DVD, so at times I have forgotten to install ffmpeg before running yt-dlp. What yt-dlp has done is save the audio and video files separately. Then, when I ran yt-dlp again to fix that problem it apparently noticed the data files had been downloaded already and just employed ffmpeg to put them together ...
